### DISCUSSION TITLE
Type error with middleware + slice pattern

### DISCUSSION THREAD

**Discussion by doliG** [pmndrs/zustand#1409] [Q&A]
Hello,

Tanks for providing us this library. I have some trouble using _slice pattern_, _middleware_ and _TypeScript_ together. Here's my code:

```ts
export const useStore = create<AuthStore & TransactionStore>()(
  devtools(
    (...args) => ({
      ...createAuthSlice(...args),
      ...createTransactionSlice(...args),
    }),
    { enabled: process.env.NODE_ENV === "development" }
  )
)
```

And got this incomprehensible typescript error.

```
Argument of type '[["zustand/devtools", never]]' is not assignable to parameter of type '[]'.
  Source has 1 element(s) but target allows only 0.
```

I also have typing problems when trying to use custom middleware

```ts
export const useStore = create<AuthStore & TransactionStore>()(
  devtools(
    withLog((...args) => ({
      ...createAuthSlice(...args),
      ...createTransactionSlice(...args),
    })),
    { enabled: process.env.NODE_ENV === "development" }
  )
)

export withLog = (config) => (set, get, api) => config(
    (...args) => {
      console.log(args)
    },
    get,
    api
  )
```

```
Parameter 'config' implicitly has an 'any' type.
Parameter 'set' implicitly has an 'any' type.
Parameter 'get' implicitly has an 'any' type.
Parameter 'api' implicitly has an 'any' type.
Rest parameter 'args' implicitly has an 'any[]' type.
```

I'm clueless on how to solve it, can you help ?

Note: in both case the code is working fine, it's just a problem of typing (got the devtools working).

**Comment by dai-shi** [comment]
You don't get type errors, if you don't use slice pattern, right?

In that case, it's the issue of typing slices. It's a bit tricky and docs could be improved to mitigate it. Please check docs and share where you found something needs to be improved.

**Reply by doliG** [reply]
So I tried without slice pattern and I have type errors. The store is not able to know the return of `(parameter) get: () => unknown` or the typing of functions. Here's a sample of the code: 

```ts
type MyStore = MyState & MyActions
type MyActions = { /* ... */ }
type MyState = { /* ... */ }


const withLog = (config) => (set, get, api) =>
  config(
    (...args) => console.log(args),
    get,
    api
  )

export const usMyStore = create<MyStore>()(
  withLog(
    devtools((set, get) => ({
      // ...
    }))
  )
)
```

It's probably be because the middleware is not typed (get, set and config have `any` type). It's very easy to improve the doc by adding a part in "middleware" or in "typescript" on how to correctly type middlewares. But I don't know how to type it, and that's currently my main blocker 🤔

Hope you can help :)

**Reply by dai-shi** [reply]
There's docs about typing middleware, but it is extremely head for beginners.

Check this out: https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#middleware-that-doesnt-change-the-store-type

For your specific case, we could do this to pass type checks: https://tsplay.dev/WJXEDN

I think it's too hard for most people and suggest not to try creating own middleware, unless you are a library author.

(We can remove the code from readme... 🤔 )

**Reply by doliG** [reply]
Indeed typing middleware seems quite difficult although it's not impossible.

Maybe the problem I'm trying to solve with middleware can be solved in a different way ?

Let me explain:

Let's say I have `A` & `B` in my store. When I change `A` or `B`, I want the store to compute `C`. And when I have `C`, I want my store to fetch some data. BUT `C` need to be debounced (`A` & `B` are inputs and I don't want to query the server at each keystroke), and also need to be refreshed every X second.

At first I've tried what you advised in #1384 but it quickly becomes a mess (very easy to make mistakes).

```ts
import create from 'zustand'

const useBearStore = create((set) => ({
  a: "",
  b: "",
  c: "",
  d: "",

  updateA: () => {
    set({ a })
    get().computeC()
    get().fetchD()
  },

  updateB: () => {
    set({ b })
    get().computeC()
    get().fetchD()
  },

  computeC: () => {
    set({ c: compute(a, b) })
  },

  fetchD: () => {
    debounce(() => {
      const d = await fetch(get().a, get().b, get().c)
      set({ d })
    }
  }
}))
```

> 💡 Note: this is a simplified example, in reality I have way more input, computed value and fetchs.

Then I tried the hook approach mentionned in #1341 , but it seems wrong because when using a store I expect to have everything inside, and hooks works outside.

```ts
import create from 'zustand'

const useBearStore = create((set) => ({
  a: "",
  b: "",

  updateA: (a) => set({ a }),
  updateB: (b) => set({ b }),
}))

const useC() => {
  const a = useStore(state => state.a)
  const b = useStore(state => state.b)
  return compute(a, b)
}

const useD() => {
  const a = useStore(state => state.a)
  const b = useStore(state => state.b)
  const c = useStore(state => state.c)
  const [d, setD] = useState({})
  useEffect(
    debounce(() => setD(fetch(get().a, get().b, get().c)), 100),// take out async logic to simplify the example
    [a, b, c, setD]
  )
  return d
}
```

> 💡 Note: this is my favorite solution so far, and it's quite a shame because it's by not using zustand ^^

Then I tried the `subscribeWithSelector` approach but I needed to use `setTimeout` because subscribe was called before store initialisation 🤦 No example needed here cause it's the same as with hooks execpt that I use `set()` in my subscriber.

Finally I tried the middleware approach, but I can't type them properly...

Honnestly I'm starting to feel like even is zustand is easy to apprehend, it's the wrong tools for my problem. In a pure react way, it's something very easy to do:

```js
const Component = () => {
  const [a, setA] = useState("")
  const [b, setB] = useState("")
  const c = compute(a, b) // can be in useMemo if expensive computation
  const [d, setD] = useState({})
  useEffect(
    debounce(() => setD(fetch(get().a, get().b, get().c)), 100),// take out async logic to simplify the example
    [a, b, c, setD]
  )
}
```

Again, thanks for your help and I'm looking forward to your answer :)

**Reply by dai-shi** [reply]
> it's the wrong tools for my problem.

I would like to hear if [jotai](https://github.com/pmndrs/jotai) fits better with your use case.

> In a pure react way, it's something very easy to do:

but, `useEffect` is the wrong tool for this case.

**Reply by doliG** [reply]
Hey, thanks again for your quick answer. Jotai looks great but I don't want to spend too much time moving from one stand to another, especially one I never tried before (I mean it's not a side project it's for my main job haha).

I checked react docs and they say it's the right place to do data fetching.
> Data fetching, setting up a subscription, and manually changing the DOM in React components are all examples of side effects
> https://reactjs.org/docs/hooks-effect.html

So can you explain why `useEffect` is the wrong tool for this case ? Thanks 🙏

**Reply by dai-shi** [reply]
https://beta.reactjs.org/learn/you-might-not-need-an-effect may help.

It doesn't explicitly say `useEffect` is wrong for data fetching.

---

https://github.com/reactjs/rfcs/pull/229

Maybe when `use` and `cache` land, they will provide a better solution for data fetching.

**Reply by doliG** [reply]
> https://beta.reactjs.org/learn/you-might-not-need-an-effect may help.
> It doesn't explicitly say `useEffect` is wrong for data fetching.

Yeah, exactly ! But previously you said:

> but, useEffect is the wrong tool for this case.

And I don't understand why. All I do in the useEffect is a debounced fetch, which is frontend common usecase 🤔 So if you can explain what's wrong with that I'm curious.

> [reactjs/rfcs#229](https://github.com/reactjs/rfcs/pull/229)
> Maybe when `use` and `cache` land, they will provide a better solution for data fetching.

Interessing 🤓 Let's wait & see

**Reply by dai-shi** [reply]
useEffect is delayed. As your use case is "debounced", it's not critical anyways. But, still useEffect isn't an ideal place to start data fetching.
I'm not sure if I can convince you or if I need to do so, but here's some tweets regarding this:
https://twitter.com/dai_shi/status/1554467083089747968
https://twitter.com/dai_shi/status/1561539728947953664

**Reply by doliG** [reply]
Indeed it's delayed, and so is async call to servers 🤔 Of course you can try to convince me, but so far I've seen no striking arguments haha

Btw I tend to agree that it's better to make call in store or using a specific lib (i.e `useQuery`), but I can't see what's wrong with putting that kind of logic in `useEffect` 🤔  Still open to discussion hehe
